notebooks_community/FuRBO/FuRBO.ipynb

```diff
@@ -0,0 +1,1011 @@
+    "                   n_candidates,\n",
+    "                   **tkwargs):\n",
+    "    '''Function to find net candidate optimum'''\n",
+    "    assert state.X.min() >= 0.0 and state.X.max() <= 1.0 and torch.all(torch.isfinite(state.Y))\n",
+    "\n",
+    "    # Initialize tensor with samples to evaluate\n",
+    "    X_next = torch.ones((state.batch_size, state.dim), **tkwargs)\n",
+    "    \n",
+    "    # Iterate over the several trust regions\n",
+    "\n",
+    "    tr_lb = state.tr_lb[0]\n",
+    "    tr_ub = state.tr_ub[0]\n",
+    "\n",
+    "    # Thompson Sampling w/ Constraints (like SCBO)\n",
+    "    pert = state.sobol.draw(n_candidates).to(**tkwargs)\n",
+    "    pert = tr_lb + (tr_ub - tr_lb) * pert\n",
+    "\n",
+    "    # Create a perturbation mask\n",
+    "    prob_perturb = min(20.0 / state.dim, 1.0)\n",
+    "    mask = torch.rand(n_candidates, state.dim, **tkwargs) <= prob_perturb\n",
+    "    ind = torch.where(mask.sum(dim=1) == 0)[0]\n",
+    "    mask[ind, torch.randint(0, state.dim - 1, size=(len(ind),), device=tkwargs['device'])] = 1\n",
+    "\n",
+    "    # Create candidate points from the perturbations and the mask\n",
+    "    X_cand = state.best_X[0].expand(n_candidates, state.dim).clone()\n",
+    "    X_cand[mask] = pert[mask]\n",
+    "    \n",
+    "    # Sample on the candidate points using Constrained Max Posterior Sampling\n",
+    "    constrained_thompson_sampling = ConstrainedMaxPosteriorSampling(\n",
+    "        model=state.Y_model, constraint_model=state.C_model, replacement=False\n",
+    "        )\n",
+    "    with torch.no_grad():\n",
+    "        X_next[0*state.batch_size:0*state.batch_size+state.batch_size, :] = constrained_thompson_sampling(X_cand, num_samples=state.batch_size)\n",
+    "        \n",
+    "    return X_next"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "024e10f4-4de9-433b-8827-c58a7177784f",
+   "metadata": {},
+   "source": [
+    "### Stopping criterion\n",
+    "\n",
+    "Define a function to detect when the maximum number of iterations is met."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 177,
+   "id": "189aae72-f033-49db-bcc8-0e791774bd76",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def stopping_criterion(state, n_iteration):\n",
+    "    '''Function to evaluate if the maximum number of allowed iterations is reached.'''\n",
+    "    if state.it_counter <= n_iteration:\n",
+    "        return False\n",
+    "    return True"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "44f15711-8503-4226-8ee0-171f26f7b8f4",
+   "metadata": {},
+   "source": [
+    "### Restart criterion\n",
+    "\n",
+    "Detect when the GPR fitting process fails to stop the optimization.curve"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 178,
+   "id": "f0d2b342-601b-4d07-9fe3-459f28ecead2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def GP_restart_criterion(state):\n",
```

Comment:
This does not seem like a good thing to include to me, especially when we only run for 100 iterations. Are you sure there is not a bug here?